### PR TITLE
Several Additions to Migration Manual

### PR DESCRIPTION
**Pull request by pdl**
Mostly things I've run into while migrating to Dancer2.

### CODE REVIEW

File: lib/Dancer2/Manual/Migration.pod [hunk 1]
```diff
@@ -500,4 +569,15 @@ delete:
...
+
+    session->destroy
+
+In Dancer2, we write:
+
+    app->destroy_session if app->has_session
+
+If you make changes to the session in an C<after> hook, your changes will
+not be written to storage, because writing sessions to storage also takes
+place in an (earlier) C<after> hook.
```
Comment (xsawyerx) [inline lib/Dancer2/Manual/Migration.pod]: I'm wondering if this is not a bug.

File: lib/Dancer2/Manual/Migration.pod [hunk 2]
```diff
@@ -298,6 +300,26 @@ function with decoded input.
 
 =back
 
+=head3 Templating engine changes
+
+Whereas in Dancer1 it was acceptable to write:
+
+    template '/foo/bar'
+
+In Dancer2, the leading slash is not permitted:
```
Comment (pdl) [inline lib/Dancer2/Manual/Migration.pod]: @xsawyerx - looks like it's both.

D1 simple: `/index.tt` treated as `APPDIR/views/index.tt`.
D1 template_toolkit: `/index.tt` treated as `APPDIR/views/index.tt`.
D2 simple: `/index.tt` treated as `APPDIR/views/index.tt`.
D2 template_toolkit: `/index.tt` behaves differently - looks like it's looking for `/index.tt`, i.e. at the filesystem root.

I guess 'not permitted' is not the right phrase. But I think we should flag this change (assuming it's deliberate), even if it's strictly not Dancer Core but the TT2 plugin.

File: lib/Dancer2/Manual/Migration.pod [hunk 3]
```diff
@@ -432,8 +454,55 @@ for these options, or the particular session engine for other supported options.
 	    is_secure: 1
         is_http_only: 1
 
+=head3 Plack Middleware
+
+In Dancer1 you could set up Plack Middleware using a C<plack_middlewares> key
+in your C<config.yml> file. Dancer2 will not respect this. Instead, invoke
+middleware in your C<app.psgi> as demonstrated in
```
Comment (pdl) [inline lib/Dancer2/Manual/Migration.pod]: @xsawyerx Yes, that's clearer, thanks.

File: lib/Dancer2/Manual/Migration.pod [hunk 4]
```diff
@@ -432,8 +454,55 @@ for these options, or the particular session engine for other supported options.
...
+Dancer2's DSL is implemented differently, and neither of these are possible
+directly. If you need this ability, you can write a package which provides
+this as follows:
+
+    package MyApp::DSL;
+    use Dancer2 appname => 'MyApp';
+
+    use Exporter qw< import >;
+
+    our @EXPORT = qw< get post ... >
```
Comment (pdl) [inline lib/Dancer2/Manual/Migration.pod]: Works for me.

File: lib/Dancer2/Manual/Migration.pod (hunk 1 above)
Comment (pdl) [inline lib/Dancer2/Manual/Migration.pod]: I had wondered that, too!

File: lib/Dancer2/Manual/Migration.pod [hunk 5]
```diff
@@ -298,6 +300,26 @@ function with decoded input.
...
+
+Whereas in Dancer1, template engines have the methods:
+
+    $template_engine->view('foo.tt')
+    $template_engine->view_exists('foo.tt')
+
+In Dancer2, you should instead write:
+
+    $template_engine->view_pathname('foo.tt')
+    $template_engine->pathname_exists($full_path)
```
Comment (pdl) [inline lib/Dancer2/Manual/Migration.pod]: I think it's probably fine to put it like that although strictly code where I found this is not writing a templating engine, it's just using TT2. Admittedly, it's wrapping `template` - but all it's really doing is checking to see if the view exists and if not, trying to find another view which does.

**Comment by pdl** [comment]
I've made most of the changes suggested. You write:

> I can see the session factory _does_ provides a `destroy` method.

I'm not sure if you mean that a) it's not an issue, b) a fix is possible, or c) something else. For me, a bare app with `session->destroy` fails.

**Comment by SysPete** [comment]
@pdl you are correct: `session->destroy` is not possible and `app->destroy_session` is the correct way. The session object cannot destroy itself since destruction is actually performed by the app's `session_engine` which the individual `session` object has no knowledge of.

For me this looks good so :+1:

**Comment by cromedome** [comment]
Thanks @pdl, this is great! :+1:

**Comment by cromedome** [comment]
Merged and I forgot to close this. Oops. Thanks again!

**Comment by pdl** [comment]
@cromedome this was closed but not merged. Perhaps you were confused by a reference from a PR in the advent-calendar repo?

**Comment by cromedome** [comment]
That could be. Sorry for the mix up!

**Comment by cromedome** [comment]
Merged (for real this time!). Thanks!

**Pull request closed without merging**